### PR TITLE
prepared sampling to only send callstacks once and hashes afterwards

### PR DESCRIPTION
**Pull request by florian-kuebler**
The SamplingProfiler now only stores the complete callstacks once per unique stack. The list of all callstacks (`m_Callstacks`) now does not contain the complete callstack data anymore. 

However, there were more callers to the `AddCallStack` method, so I decided to temporarily keep it.

Also the `EventBuffer::AddCallstackEvent` method does not take a callstack anymore.

With this, it will be quite simple to not send the complete callstacks on live mode anymore.

### CODE REVIEW

File: OrbitCore/SamplingProfiler.cpp [hunk 1]
```diff
@@ -254,14 +290,12 @@ void SamplingProfiler::ProcessSamples()
...
-    for( CallStack & callstack : m_Callstacks )
+    for( HashedCallStack & callstack : m_Callstacks )
     {
-        callstack.Hash();
-        auto it = m_UniqueCallstacks.find( callstack.m_Hash );
-        if( it == m_UniqueCallstacks.end() )
-        {
-            m_UniqueCallstacks[callstack.m_Hash] = std::make_shared<CallStack>(callstack);
-        }
+        assert( HasCallStack(callstack.m_Hash) );
```
Comment (pierricgimmig) [inline OrbitCore/SamplingProfiler.cpp]: We should log an error message instead of asserting.

File: OrbitCore/SamplingProfiler.cpp [hunk 2]
```diff
@@ -145,6 +145,42 @@ std::multimap<int, CallstackID> SamplingProfiler::GetCallStacksFromAddress( uint
...
+    hashedCS.m_Hash = hash;
+    hashedCS.m_Depth = a_CallStack.m_Depth;
+    hashedCS.m_ThreadId = a_CallStack.m_ThreadId;
+    AddHashedCallStack ( hashedCS );
+}
+
+//-----------------------------------------------------------------------------
+void SamplingProfiler::AddHashedCallStack( HashedCallStack & a_CallStack )
+{
+    assert( m_State == Sampling );
```
Comment (pierricgimmig) [inline OrbitCore/SamplingProfiler.cpp]: please convert asserts to logs

File: OrbitCore/SamplingProfiler.cpp [hunk 3]
```diff
@@ -254,14 +290,12 @@ void SamplingProfiler::ProcessSamples()
...
+    for( HashedCallStack & callstack : m_Callstacks )
     {
-        callstack.Hash();
-        auto it = m_UniqueCallstacks.find( callstack.m_Hash );
-        if( it == m_UniqueCallstacks.end() )
-        {
-            m_UniqueCallstacks[callstack.m_Hash] = std::make_shared<CallStack>(callstack);
-        }
+        assert( HasCallStack(callstack.m_Hash) );
+        //{
```
Comment (pierricgimmig) [inline OrbitCore/SamplingProfiler.cpp]: Is this commented on purpose?  Do we need to keep the comments?

File: OrbitCore/SamplingProfiler.h [hunk 4]
```diff
@@ -95,8 +95,12 @@ class SamplingProfiler
     float GetSampleTimeTotal() const { return m_SampleTimeSeconds; }
     bool  ShouldStop();
     void FireDoneProcessingCallbacks();
-    void AddCallStack( CallStack & a_CallStack ) { if( m_State == Sampling ) m_Callstacks.push_back( a_CallStack ); }
+    // TODO: remove this method in the future.
+    void AddCallStack( CallStack & a_CallStack );
+    void AddHashedCallStack( HashedCallStack & a_CallStack );
+    void AddUniqueCallStack( CallStack & a_CallStack );
     const std::shared_ptr<CallStack> GetCallStack( CallstackID a_ID ) { return m_UniqueCallstacks[a_ID]; }
+    bool HasCallStack( CallstackID a_ID );
```
Comment (pierricgimmig) [inline OrbitCore/SamplingProfiler.h]: This could be "inline"

**Comment by florian-kuebler** [comment]
Thanks for the review. I addressed all your comments.

**Pull request merged**